documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,488 @@
+
+- A new popularity metric is added (for a new or existing provider)
+- The popularity constants are recalculated
+
+In both cases, we need to refresh the `<media>_popularity_constants` view (in
+order to recalculate constants). Note that once the constants have been updated,
+_the provider DAGs will immediately pick up the new constant and begin using it
+at ingestion_. All that remains is to update the scores for historical data.
+
+**Note**: It is true that our reingestion DAGs will also immediately begin
+updating the standardized scores of historical data using the new constant.
+However, this process is slow and some reingestion DAGs are currently running on
+sparse schedules. I think it is still worth it to explicitly updates the scores
+in this DAG, especially because that will allow us to use the DAG to do the
+initial backfill.
+
+As with our data refresh DAGs, we will use a factory to generate a
+`popularity_refresh` DAG for each media type. The DAG will do the following:
+
+1. Update the `<media>_popularity_metrics` table to include any newly added
+   metrics.
+1. Refresh the `<media>_popularity_constants` view to recalculate the popularity
+   constants.
+   1. This is done `CONCURRENTLY` so that provider DAGs can continue reading
+      from the view while it updates.
+1. For each unique `provider` in the `<media>_popularity_constants` view,
+   generate a `refresh_<provider>_scores` task. The task will run an `UPDATE` of
+   the `standardized_popularity` on all records matching that provider which
+   were last updated before the task began.
+   1. We may consider running the `refresh_<provider>_scores` tasks in parallel
+      to speed up the update.
+   1. Optionally, we can hard code a `SKIPLIST` of providers that are present in
+      the `<media>_popularity_constants` view, but for which we do not want to
+      create a refresh task. We currently have some providers (Nappy, Rawpixel,
+      Stocksnap) that support popularity data but are not dated, meaning scores
+      for **all** of their records will be updated the next time the DAG runs.
+      _Note that some of these DAGs are on a `@monthly` schedule however, which
+      means skipping them in this DAG could result in delayed recalculation
+      time._
+1. Report to Slack when the scores have finished updating.
+
+Some notes:
+
+- The DAG will initially run on a `None` schedule, while we get a sense of how
+  long it will take to complete. Thereafter we will enable it on a `@monthly`
+  schedule if possible, or quarterly if it proves too long-running.
+- This DAG can run completely independently of the data refresh. The data
+  refresh will always have access to the most recently ingested records, with
+  standardized popularity. They will at worst be calculated using slightly
+  outdated constants, which is the case today in production.
+
+#### Special Considerations: Avoiding Deadlocks and Timeouts
+
```

Comment:
I wonder if it would be possible to handle conflicts at a higher level. Just thinking out loud... would it be possible to separate out old records from dated DAGs with popularity data, from the new/current run only data? Or is it that the whole point of dated DAGs is to still do a full refresh, just using dates to break it into manageable chunks? The smaller commit strategy is very cool, but the switching costs could be significant, so I guess I'm wondering if there is a way to limit the scope of Step 5 below.

Reply:
I went down some similar lines of thought while working on this, but unfortunately I didn't see a way. We want to accommodate both the insertion of new records and the updating of old records by provider DAGs. Those updates of old records are crucial in part because they include updates to raw popularity scores, so we need this information ASAP. Moreover, it's not just dated DAGs that have popularity data -- some non-dated DAGs, which reingest all records on each run, have popularity scores. 

Since we don't want to prevent the provider DAGs from inserting/updating any of their data, I don't see any way to prevent them from getting stuck awaiting the popularity refresh without batching it (although I'm very open to ideas!). There is definitely a significant cost to the batching, but since we're tolerant to the popularity refresh being long-running as long as it does not block our other processes, I think this is acceptable.

Reply:
Hmmm... Sorry if this is too late, please feel free to ignore. But I guess my suggestion was about limiting the records included in the popularity refresh so that there isn't a conflict. For providers with non-dated DAGs, every ingestion is a full refresh, right? So their popularity stuff would be handled at ingestion and (I think?) there wouldn't be a need for a popularity refresh at all. (Unless they run so rarely that the ingestion couldn't take care of it, but in that case, a conflict would be pretty unlikely, right?) I guess maybe I was assuming that it would be possible in the dated dags, to divide records into two groups -- those where we need to update (or insert) _both_ popularity calculations _and_ other data (handled by the regular provider dags), and those where we _only_ want to update the popularity calculation (handled by the refresh). And that maybe some combination of the last update field, DAG type / schedules / statuses from airflow, could create a useful filter for the popularity refresh that would allow it to complete more quickly while avoiding records likely to be handled by the regular dag.

Reply:
Inserts would definitely be handled by the regular provider dag, and wouldn't lock the whole table, so they don't seem like a concern here. Right? And I still think doing one provider at a time would be useful, maybe building a temp table and [merging](https://www.postgresql.org/docs/current/sql-merge.html) it in could be more performant for the popularity only updates.

Reply:
OK, yeah, I just started to look in airflow, and made [this super basic spreadsheet](https://docs.google.com/spreadsheets/d/1yzmFbs2L4yiu3yCqzWNNIPb_uUadryetMaNo_67mTOg/edit#gid=864570427). So, I think this is where we are, right?

- Most providers don't have popularity data, so it's a non-issue for them.
- Of those that do, four of them (Nappy, Jamendo, Rawpixel, and Stocksnap) are relatively small/fast, non-dated, _monthly_ dags. So it would be pretty straightforward to just run the refresh for each of those providers individually by building a provider temp table with just the UUID and popularity calculation using a select from the main table, and then doing a merge back into the main table for the update. And to avoid conflicts, we'd just need to pick a day when the regular dag wasn't running.
- That leaves flickr and wikimedia. And they're tricky, because they have daily, dated, dags, that processed <50,000 records in <15 minutes the last time they ran, but I guess we don't know how many of those were inserts and how many were updates. The inserts don't (or shouldn't) matter for refresh purposes. But let's say that there were 1,000 updates in each run, and that we have no way of knowing where those 1,000 will fall in the (guessing here) 300,000,000 records we have from each of those providers. So, could we instead, do a similar process as for the smaller providers described above, but further divide these providers by `updated_on`, taking a month at a time or something? If that won't work because of the recency of implementing dated dags (i.e. the vast majority have the date of  last run prior to dated dags), then maybe we'd use `created_on` in addition to get a reasonably sized temp table. Start at the beginning and work forward. We could pause the dag when the regular daily run is active, but we wouldn't have to if there was a filter in the merge statement to say something like "if this has been updated since we started the refresh, skip it". That's assuming that the refresh starts after the ongoing provider runs are already calculating popularity, which I think was the order, right?

Reply:
If `updated_on` + `created_on`, still leaves too big of a partition, maybe there is a further partition using the first n characters of the uuid or something?

Reply:
Hm, I'm not sure I'm following. I think I generally agree but I'm not sure what changes you mean to the proposed solution 🤔 

**Non-dated DAGs**
I pointed out [here](https://github.com/WordPress/openverse/pull/1894/files#diff-9d0dc40138273796366c2301a730b7da43513f37364fdc77bc2835b7aff4aa14R220) that non-dated DAGs will get their popularity scores updated the next time their DAG runs, as you mentioned. I suggested having a `SKIPLIST` for those providers and letting them refresh ‘naturally’, as one option. I’m not convinced that we should, though: my reasoning for not doing so is that the popularity refresh could feasibly complete just moments after the provider DAG, and then popularity data for that provider will be ‘stale’ for an entire month. If we really want to avoid doing popularity refreshes for them we could get away with it but I don’t see any real benefit to not? Particularly if the providers refreshes happen in parallel.

> So it would be pretty straightforward to just run the refresh for each of those providers individually by building a provider temp table with just the UUID and popularity calculation using a select from the main table, and then doing a merge back into the main table for the update. And to avoid conflicts, we'd just need to pick a day when the regular dag wasn't running.

If we’re for sure running it when the regular DAG is not running I don’t think you need a temp table at all, you could just do the update. But I also would not _rely_ on offset default schedules to guarantee that these DAGs don’t run at the same time, especially since we need to make sure this works for any new providers that are added in the future, which may have schedules and runtimes we haven’t considered. We could add in sensors to make sure the provider DAG isn't running, but that's additional complexity -- and I'd prefer for the provider DAGs to _never_ be blocked on popularity refreshes.

Using the batched update is slow, but incremental updates are committed ASAP (which is not the case if we wait for the regular DAG run to naturally refresh data) and we’re guaranteed to have no conflicts, no matter what the DAG’s schedule is, or if a backfill is running or a DAG is being retried on an atypical day/some other unforeseen problem. Plus, we don't have to maintain a separate process for updating dated vs non-dated DAGs.

**Dated DAGs**
I pointed out [here](https://github.com/WordPress/openverse/pull/1894/files#diff-9d0dc40138273796366c2301a730b7da43513f37364fdc77bc2835b7aff4aa14R198) that our existing reingestion workflows do, in theory, eventually refresh all records for our dated DAGs, so if we really wanted to we could also rely on these to naturally refresh popularity data. Unfortunately at the moment:
* these are in many cases flaky
* Flickr’s is actually disabled due to rate limiting concerns
* many are on sparser schedules than we would like (running only `weekly` instead of `daily`) due to the flakiness, which means they are not reingesting enough data to reliably and fully refresh all records

Reingestion DAGs are always, 100% of the time doing updates rather than inserts that risk conflict with the popularity refresh. We don’t want to turn off or interfere with the reignestion workflows, because they update more than just popularity scores. At the moment you could maybe schedule around them because they are on weekly schedules, but we intend to change them to daily schedules, which will make that impossible. Regular dated workflows do a mix of inserts and updates and as you point out, we don’t have any way of guessing which rows are going to be updated.

>  If that won't work because of the recency of implementing dated dags (i.e. the vast majority have the date of last run prior to dated dags), then maybe we'd use created_on in addition to get a reasonably sized temp table. Start at the beginning and work forward. We could pause the dag when the regular daily run is active, but we wouldn't have to if there was a filter in the merge statement to say something like "if this has been updated since we started the refresh, skip it".

This sounds really similar to me to the proposed solution. “If this has been updated since we started the refresh, skip it” is essentially what I’m accomplishing by using `SKIP LOCKED`. The batching is to prevent the provider DAG from hanging/waiting on the refresh, and I still do not see any way around that at all. I think you're saying theh same thing when you talk about partitions, is that right? Or do you see a way to avoid batches altogether?

When I build the temp table that is used to generate batches, I already [only attempt to update records](https://github.com/WordPress/openverse/pull/1894/files#diff-9d0dc40138273796366c2301a730b7da43513f37364fdc77bc2835b7aff4aa14R302) that were last updated before the constants were recalculated. It is true that some of those records will end up being updated by the reingestion workflows in the background, after that temp table is built, so the `last_updated` filter isn't perfect.

One nice optimization would be to add one more check of the `last_updated` date to the update for each batch. Ie:

```pgsql
  -- Loop over `rows_to_update` 10k rows at a time, updating scores.
  -- Avoid deadlock using SKIP LOCKED
  FOR i IN min_row..max_row BY rows_per_loop LOOP
   UPDATE image
   SET standardized_popularity = standardized_image_popularity(
       image.provider,
       image.meta_data
   )
   WHERE identifier in (
       SELECT identifier FROM rows_to_update
       WHERE 
            row_id >= i
            AND row_id < i+rows_per_loop
            AND updated_on < max_timestamp
       FOR UPDATE SKIP LOCKED
   );
   -- COMMIT this batch
   COMMIT;
```

Is that all that you're suggesting? I'm sorry if I'm totally off mark 😅

Reply:
Here's what I think is different, and I haven't tested it, but I guess I'm thinking about doing the calculation in the table build, so it could be something more like this:

```
create temp_table as (
    select 
        identifier, 
        now() as cut_off_timestamp, 
        standardized_image_popularity( image.provider, image.meta_data ) as standardized_popularity
    from image
    where provider='flickr' and [some constants calculated in a loop in python to divide the table into ~100,000 record chunks, using updated_on, created_on, and maybe a substring of identifier]
);
commit;
merge into image using temp_table on (identifier)
when matched and (image.updated_on < temp_table.cut_off_timestamp)
set image.standardized_popularity = temp_table.standardized_popularity
;
commit;
```

Doing it this way would avoid the issue of updating records that had more recently been updated (by whatever job, with the date filter), might be faster than using row_id (at least that's the lesson learned from improving iNaturalist performance, it would depend on indexing on the image table, but I don't think it would be worse than this), and hopefully makes the update freeze itself a lot faster by moving the calculation out of the update step and into the table build. And because the chunks are meaningfully defined / sorted, there is less risk of traffic jams for new records, right?

Reply:
Thanks for writing out the queries, that really helps! Sorry to be long-winded, but I’m going to re-state my understanding of things to make sure I understand, and you can correct me where I’m wrong please 😄 

So a problem with the approach lined out in the IP is that we build the whole temp table of `records_to_update` at the beginning and then split it into batches. By the time we get around to processing any given batch, many of the records in that batch might already have been updated by another DAG, so it is a waste of time to update it again.

I [tried to address this](https://github.com/WordPress/openverse/pull/1894#discussion_r1187900145) by adding an additional filter to the `update` query for each batch, to filter out any records that have since been updated. So while each batch will have 10k candidate records in it, many of them might get filtered out and it will process potentially much faster than expected.

The alternative you’ve proposed, as I understand it (please correct me), is to instead first programmatically identify a series of constants that can be used in a query to divide the records into rough batches. The simplest version of this would be a series of timestamps `[a, b, c, … n]` such that there are about 10k records whose `updated_on` date is between each interval `(a, b)`, `(b, c)`, and so on. Then for each of these intervals, we create a temp table for just that batch, and we calculate the popularity score while building this table. This is then merged into the `image` table, rather than doing an update.

So a comparison/contrast:
* preventing updates to records that have been more recently updated by a provider DAG:
    * Both ideas use a filter on some `cut_off_timestamp` in the update/merge query to prevent unnecessary updates
* length of time the table is locked
    * Since your idea does the calculation during the table build, the `image` table will not be locked for as long. But since we’re forced to process in batches anyway, I don’t think we need to be particularly concerned: provider DAGs will never have to wait on it long
* “because the chunks are meaningfully defined / sorted, there is less risk of traffic jams for new records”
    * This part I’m not sure I understand. In both cases we try to avoid updating records with a recent `updated_on` date, which is really the only factor in collisions. I still think we need to use `SKIP LOCKED`, but that should work fine in either case.

You mentioned that it might be faster though than using `row_id`, which I’m interested in investigating. It would require a lot more work and complexity to actually create the batches (if I’m understanding what you meant about using “constants calculated in a loop in python to divide the table”), right? I planned on using `row_id` because I thought that was supposed to be much faster than sorting and selecting with very large `OFFSET`s, which is the first thing I thought of for calculating those constants — but I could be totally wrong, or else maybe you had a different idea for how to calculate them?

I don't have a good sense of how we would reliably make those partitions. It _sounds_ like it would be much more complex and potentially tricky (again, please correct me if this assumption is wrong!), so I'm a little hesitant to do so if we're not absolutely certain that it will be _significantly_ faster, especially since the popularity refresh is a process that we actually don't mind being very slow. We may find that the `row_id` approach takes an extremely long time (>month), in which case it would definitely make sense to prioritize further optimizations.